From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IORA Build System <build@iora.local>
Date: Sat, 3 May 2026 00:00:00 +0000
Subject: [PATCH] Fix GCC 15/C23: bool/true are keywords

C23 makes bool, true, false reserved keywords. Go 1.4 bootstrap uses
typedef int bool and uses 'true' as a parameter name.

Guard the typedef + rename the parameter to avoid C23 conflicts.

--- a/include/u.h
+++ b/include/u.h
@@ -49,7 +49,9 @@
 #define nil			((void*)0)
 #endif
 
+#if __STDC_VERSION__ < 202311L
 typedef	int		bool;
+#endif
 typedef	struct	Algn	Algn;
 typedef	struct	Basic	Basic;
 typedef	struct	Bphist	Bphist;
--- a/src/cmd/6c/cgen.c
+++ b/src/cmd/6c/cgen.c
@@ -1223,7 +1223,7 @@
 
 void
-boolgen(Node *n, int true, Node *nn)
+boolgen(Node *n, int istrue, Node *nn)
 {
 	int o;
 	Prog *p1, *p2, *p3;
@@ -1231,7 +1231,7 @@
 	int32 curs;
 
 	if(debug['g']) {
-		print("boolgen %d\n", true);
+		print("boolgen %d\n", istrue);
 		prtree(nn, "boolgen lhs");
 		prtree(n, "boolgen");
 	}
@@ -1242,7 +1242,7 @@
 	default:
 		o = ONE;
 
-		if(true)
+		if(istrue)
 			o = OEQ;
 
 		/* bad, 13 is address of external that becomes constant */
@@ -1269,7 +1269,7 @@
 
 	case OCONST:
 		o = vconst(n);
-		if(!true)
+		if(!istrue)
 			o = !o;
 		gbranch(OGOTO);
 		if(o) {
@@ -1297,7 +1297,7 @@
 
 	case OCOMMA:
 		cgen(l, Z);
-		boolgen(r, true, nn);
+		boolgen(r, istrue, nn);
 		break;
 
 	case OANDAND:
@@ -1453,5 +1453,5 @@
 void
 bool64(Node *n)
 {
-	boolgen(n, 1, Z);
+	boolgen(n, 1, Z);
 }
--- a/src/cmd/6g/cgen.c
+++ b/src/cmd/6g/cgen.c
@@ -1190,7 +1190,7 @@
  */
 void
 bgen(Node *n, int true_)
-{
 {
 	Type *et;
 	Node n1, n2, n3, *nl, *nr;
@@ -1332,7 +1332,7 @@
 void
 boolgen(Node *n, int true_)
 {
-	boolgen(n, true_, Z);
+	boolgen(n, 1, Z);
 }
